lib/SILOptimizer/Transforms/CopyForwarding.cpp

```diff
@@ -1324,10 +1325,244 @@ class CopyForwardingPass : public SILFunctionTransform
+/// generates). Instead we guarantee that all normal uses are within the block
+/// of the temporary and look for the last use, which effectively ends the
+/// lifetime.
+bool TempRValueOptPass::checkNoSourceModification(CopyAddrInst *copyInst,
+                     const llvm::SmallPtrSetImpl<SILInstruction *> &useInsts) {
+  unsigned NumLoadsFound = 0;
+  auto iter = std::next(copyInst->getIterator());
+  // We already checked that the useful lifetime of the temporary ends in
+  // the initialization block.
+  auto iterEnd = copyInst->getParent()->end();
+  for (; iter != iterEnd; ++iter) {
+    SILInstruction *I = &*iter;
+
+    if (useInsts.count(I))
+      NumLoadsFound++;
+
+    // If this is the last use of the temp and modifies the source it is ok.
+    if (NumLoadsFound == useInsts.size())
+      return true;
+
+    if (AA->mayWriteToMemory(I, copyInst->getSrc())) {
+      DEBUG(llvm::dbgs() << "  Source modified by" << *iter);
+      return false;
+    }
+  }
+  // For some reason, not all normal uses have been seen between the copy and
+  // the end of the initialization block. We should never reach here.
+  return false;
+}
+
+/// Tries to perform the temporary rvalue copy elimination for \p copyInst
+bool TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
+  if (copyInst->isTakeOfSrc() || !copyInst->isInitializationOfDest())
+    return false;
+
+  AllocStackInst *tempObj = dyn_cast<AllocStackInst>(copyInst->getDest());
+  if (!tempObj)
+    return false;
+
+  assert(tempObj != copyInst->getSrc() &&
+           "can't initialize temporary with itself");
+
+  // Scan all uses of the temporary storage (tempObj) to verify they all refer
+  // to the value initialized by this copy. It is sufficient to check that the
+  // only users that modify memory are the copy_addr [initialization] and
+  // destroy_addr.
+  llvm::SmallPtrSet<SILInstruction *, 8> loadInsts;
+  for (auto *useOper : tempObj->getUses()) {
+    SILInstruction *user = useOper->getUser();
+
+    if (user == copyInst)
+      continue;
+
+    // Destroys and deallocations are allowed to be in a different block.
+    if (isa<DestroyAddrInst>(user) || isa<DeallocStackInst>(user))
+      continue;
+
+    if (!collectLoads(user, tempObj, loadInsts))
+      return false;
+  }
+
+  // Check if the source is modified within the lifetime of the temporary.
+  if (!checkNoSourceModification(copyInst, loadInsts))
+    return false;
+
+  DEBUG(llvm::dbgs() << "  Success: replace temp" << *tempObj);
+
+  // Do a "replaceAllUses" by either deleting the users or replacing them with
+  // the source address. Note: we must not delete the original copyInst because
+  // it would crash the instruction iteration in run(). Instead the copyInst
+  // gets identical Src and Dest operands.
+  while (!tempObj->use_empty()) {
+    Operand *use = *tempObj->use_begin();
+    SILInstruction *user = use->getUser();
+    switch (user->getKind()) {
```

Comment:
I was really dead-set on not rewriting replace-all-uses, but this is really a better approach.